File: Word Search/Word_Search.py
# 228ms 85.47%
class Solution:
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        m, n = len(board), len(board[0])
        if m is 0 or n is 0 or not word:
            return False

        def dfs(temp_word, start_i, start_j, path):
            if board[start_i][start_j] == word[len(temp_word)]:
                temp_word += board[start_i][start_j]
                if temp_word == word:
                    return True
                # 从上下左右四个方向分别进行搜索，这里一定要注意列表的浅拷贝问题
                if start_i + 1 < m and (start_i + 1, start_j) not in path and dfs(temp_word, start_i + 1, start_j, path | {(start_i, start_j)}):
                    return True
                if start_i - 1 >= 0 and (start_i - 1, start_j) not in path and dfs(temp_word, start_i - 1, start_j, path | {(start_i, start_j)}):
                    return True
                if start_j + 1 < n and (start_i, start_j + 1) not in path and dfs(temp_word, start_i, start_j + 1, path | {(start_i, start_j)}):
                    return True
                if start_j - 1 >= 0 and (start_i, start_j - 1) not in path and dfs(temp_word, start_i, start_j - 1, path | {(start_i, start_j)}):
                    return True
            else:
                return False
                    
        for i in range(m):
            for j in range(n):
                if dfs('', i, j, set()):
                    return True
        return False

File: Word Search/Word_Search.cpp
// 典型的回溯
// Runtime: 28 ms, faster than 75.07% of C++ online submissions for Word Search.
// Memory Usage: 12.5 MB, less than 52.94% of C++ online submissions for Word Search.

class Solution
{
public:
	bool exist(vector<vector<char>>& matrix, string word)
	{
		// 边界条件处理
		if (word.size() == 0) return false;

		int row = matrix.size();
		if (row == 0) return false;

		int col = matrix[0].size();
		if (col == 0) return false;

		// 循环查找
		vector<vector<bool>> visited(row, vector<bool>(col, false));
		for (int i = 0; i < row; ++i)
		{
			for (int j = 0; j < col; ++j)
			{
				if (matrix[i][j] == word[0])
				{
					visited[i][j] = true;

					if (dfs(matrix, word, visited, i, j, 1))
						return true;

					visited[i][j] = false;
				}
			}
		}
		return false;
	}
private:
	bool dfs(const vector<vector<char>>& matrix, const string& word, vector<vector<bool>>& visited, int starti, int startj, int start)
	{
		if (start == word.length())
			return true;
		else
		{
			// 向上寻找
			if (starti - 1 >= 0 && !visited[starti - 1][startj] && matrix[starti - 1][startj] == word[start])
			{
				visited[starti - 1][startj] = true;
				if (dfs(matrix, word, visited, starti - 1, startj, start + 1))
					return true;
				visited[starti - 1][startj] = false;
			}

			// 向下寻找
			if (starti + 1 < matrix.size() && !visited[starti + 1][startj] && matrix[starti + 1][startj] == word[start])
			{
				visited[starti + 1][startj] = true;
				if (dfs(matrix, word, visited, starti + 1, startj, start + 1))
					return true;
				visited[starti + 1][startj] = false;
			}

			// 向左寻找
			if (startj - 1 >= 0 && !visited[starti][startj - 1] && matrix[starti][startj - 1] == word[start])
			{
				visited[starti][startj - 1] = true;
				if (dfs(matrix, word, visited, starti, startj - 1, start + 1))
					return true;
				visited[starti][startj - 1] = false;
			}

			// 向右寻找
			if (startj + 1 < matrix[0].size() && !visited[starti][startj + 1] && matrix[starti][startj + 1] == word[start])
			{
				visited[starti][startj + 1] = true;
				if (dfs(matrix, word, visited, starti, startj + 1, start + 1))
					return true;
				visited[starti][startj + 1] = false;
			}

			return false;
		}
	}
};